Fluentd for Audit Logging (#2853)
antrea-io/antrea: docs/cookbooks/fluentd/README.md

```

### Step 3: Configure Custom Fluentd Plugins

The architecture of Fluentd is a pipeline from input-> parser-> buffer->
output-> formatter, many of these are plugins that could be configured to
fit users’ different use cases.

To specify custom input plugins and parsers, modify `./resources/kubernetes.conf`
and create a ConfigMap with the following command. Later, direct Fluentd
DaemonSet to refer to that ConfigMap. To see more variations of custom
configuration, refer to
[Fluentd inputs](https://docs.fluentbit.io/manual/pipeline/inputs).
This cookbook uses the [tail](https://docs.fluentbit.io/manual/pipeline/inputs/tail)
input plugin to monitor the audit logging files for Antrea-native policies
on every K8s Node.

```bash
kubectl create configmap fluentd-conf --from-file=docs/cookbooks/fluentd/resources/kubernetes.conf --namespace=kube-logging
```

### Step 4: Deploy Fluentd DeamonSet

Fluentd deployment includes RBAC and DeamonSet. Fluentd will collect logs
from cluster components, so permissions need to be granted first through
RBAC. Create a ServiceAccount, and grant permissions to read, list and
watch Pods in ClusterRole. Then, use ClusterRoleBinding to link the two

> resources.

Reviewer: the phrasing of this seems to imply that the user needs to create those resources manually, when my understanding is that they are defined in `fluentd.yml`

Maybe rephrase to:
> In `fluentd.yml`, we create a ServiceAccount, and use a ClusterRole and a ClusterRoleBinding to grant it permissions to read, list and watch Pods in cluster scope.

Then the last sentence is no longer necessary

> In the DeamonSet configuration, specify Elasticsearch host, port and scheme,

Reviewer: s/DeamonSet/DaemonSet

> as the Elasticsearch output plugin requires.

Reviewer: s/as the Elasticsearch output plugin requires./as they are required by the Elasticsearch output plugin

# Using Antrea with Fluentd

This guide will describe how to use Project Antrea with

> [Fluentd](https://github.com/fluent/fluentd-kubernetes-daemonset), in order to efficiently audit logging.

Reviewer 2: ```suggestion
[Fluentd](https://github.com/fluent/fluentd-kubernetes-daemonset), in order for efficient audit logging.
```
or something like that.. perhapss Antonin has a better suggestion

All the required software will be deployed using YAML manifests, and the
corresponding container images will be downloaded from public registries.

## Practical steps

### Step 1: Deploying Antrea

For detailed information on the Antrea requirements and instructions on how to
deploy Antrea, please refer to
[getting-started.md](../../getting-started.md). To deploy the latest version of
Antrea, use:

```bash
kubectl apply -f https://raw.githubusercontent.com/antrea-io/antrea/main/build/yamls/antrea.yml
```

You may also choose a [released Antrea
version](https://github.com/antrea-io/antrea/releases).

### Step 2: Deploy Elasticsearch and Kibana Dashboard

Fluentd supports multiple [output plugins](https://www.fluentd.org/plugins).
Details will be discussed later, for default setting, log records are collected by Fluentd DeamonSet and sent to Elasticsearch.
Then use Kibana Dashboard to visualize the data. The YAML file for deployment is included in `build` directory, use:

```bash
kubectl apply -f docs/cookbooks/build/kibana-elasticsearch.yml
```

### Step 3: Grant Permissions to Fluentd

Fluentd needs to collect logs from cluster components, so permissions needs to be granted first.

> All Fluentd related resources will be deployed in `kube-logging` namespace, together with Elasticsearch and Kibana.

Reviewer 2: ```suggestion
All Fluentd related resources will be deployed in `kube-logging` Namespace, together with Elasticsearch and Kibana.
```

> Create a ServiceAccount for Fluentd DaemonSet, in ClusterRole grant permission to read, list and watch pods.

Reviewer 2: ```suggestion
Create a ServiceAccount for Fluentd DaemonSet, in ClusterRole grant permission to read, list and watch Pods.
```

refer to [Fluentd inputs](https://docs.fluentbit.io/manual/pipeline/inputs).
This documentation uses file tailing for audit logging Antrea-native network polices at `/var/log/antrea/networkpolicy`.

```bash
kubectl create configmap fluentd-conf --from-file=docs/cookbooks/build/kubernetes.conf --namespace=kube-logging
```

### Step 5: Deploy Fluentd DeamonSet

With current [Fluentd repository](https://github.com/fluent/fluentd-kubernetes-daemonset), the output sources are specified in `fluent.conf` used by the image.
To change output sources, choose a different image and specify in `build/fluentd.yml`, note to match the version with deployed elasticsearch.

```bash
kubectl apply -f docs/cookbooks/build/fluentd.yml
```

### Step 6: Visualize with Kibana Dashboard

Navigate to `http://[NodeIP]: 30007` and create an index pattern with "fluentd-*".
Go to `http://[NodeIP]: 30007/app/kibana#/discover` to see the results as below.

<img src="./assets/audit_logging.png" width="900" alt="Audit Logging Visualization">

## Email Alerting

Kibana dashboard supports creating alerts with the logs in this [guide](https://www.elastic.co/guide/en/kibana/current/alerting-getting-started.html).

> In case of using different output sources and td-agent (the stable version of Fluentd),

Reviewer 2: this statement looks incomplete? or requires restructuring
